File: hardware_drivers.py
#!/usr/bin/env python3
"""
Hardware Drivers for Automated Planter
Raspberry Pi GPIO control for sensors and actuators
"""

import time
import json
import logging
from typing import Dict, List, Optional, Tuple
from datetime import datetime

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# GPIO pins configuration (based on your hardware setup)
GPIO_CONFIG = {
    # Sensors
    "dht22_pin": 4,           # DHT22 temperature/humidity sensor
    "soil_moisture_pin": 18,  # Capacitive soil moisture sensor (analog)
    "light_sensor_sda": 2,    # Digital light sensor I2C SDA
    "light_sensor_scl": 3,    # Digital light sensor I2C SCL
    "water_level_pins": [5, 6, 7],  # Contact water level sensors (3 positions)
    
    # Actuators
    "pump1_pin": 23,          # Peristaltic pump 1
    "pump2_pin": 24,          # Peristaltic pump 2
    "pump_enable_pin": 25,    # MOSFET enable for pumps
    
    # Status LEDs
    "status_led": 12,
    "warning_led": 13
}

class HardwareInterface:
    """Main hardware interface class for Raspberry Pi"""
    
    def __init__(self, simulation_mode: bool = False):
        """
        Initialize hardware interface
        Args:
            simulation_mode: If True, use simulated data instead of real hardware
        """
        self.simulation_mode = simulation_mode
        self.gpio_initialized = False
        
        if not simulation_mode:
            try:
                import RPi.GPIO as GPIO
                self.GPIO = GPIO
                self._init_gpio()
                logger.info("Hardware interface initialized successfully")
            except ImportError:
                logger.warning("RPi.GPIO not available, using simulation mode")
                self.simulation_mode = True
        
        # Initialize sensor data
        self.last_sensor_readings = {}
        self.sensor_history = []
        
        # Initialize pumps
        self.pump_status = {
            "pump1": False,
            "pump2": False,
            "last_watered": None
        }
    
    def _init_gpio(self):
        """Initialize GPIO pins"""
        if self.simulation_mode:
            return
            
        self.GPIO.setmode(self.GPIO.BCM)
        self.GPIO.setwarnings(False)
        
        # Set up output pins (pumps, LEDs)
        output_pins = [
            GPIO_CONFIG["pump1_pin"],
            GPIO_CONFIG["pump2_pin"], 
            GPIO_CONFIG["pump_enable_pin"],
            GPIO_CONFIG["status_led"],
            GPIO_CONFIG["warning_led"]
        ]
        
        for pin in output_pins:
            self.GPIO.setup(pin, self.GPIO.OUT)
            self.GPIO.output(pin, self.GPIO.LOW)
        
        # Set up input pins (sensors)
        input_pins = [
            GPIO_CONFIG["dht22_pin"],
            GPIO_CONFIG["soil_moisture_pin"]
        ] + GPIO_CONFIG["water_level_pins"]
        
        for pin in input_pins:
            self.GPIO.setup(pin, self.GPIO.IN, pull_up_down=self.GPIO.PUD_UP)
        
        self.gpio_initialized = True
        logger.info("GPIO pins initialized")
    
    def read_dht22(self) -> Tuple[float, float]:
        """
        Read temperature and humidity from DHT22 sensor
        Returns:
            Tuple of (temperature_celsius, humidity_percent)
        """
        if self.simulation_mode:
            # Simulate realistic values
            import random
            temp = round(20 + random.uniform(-3, 8), 1)  # 17-28°C
            humidity = round(40 + random.uniform(-10, 20), 1)  # 30-60%
            return temp, humidity
        
        try:
            import Adafruit_DHT
            humidity, temperature = Adafruit_DHT.read_retry(
                Adafruit_DHT.DHT22, 
                GPIO_CONFIG["dht22_pin"]
            )
            
            if humidity is not None and temperature is not None:
                return round(temperature, 1), round(humidity, 1)
            else:
                logger.warning("DHT22 sensor read failed")
                return None, None
                
        except ImportError:
            logger.error("Adafruit_DHT library not installed")
            return None, None
        except Exception as e:
            logger.error(f"DHT22 read error: {e}")
            return None, None
    
    def read_soil_moisture(self) -> float:
        """
        Read soil moisture from capacitive sensor
        Returns:
            Soil moisture percentage (0-100%)
        """
        if self.simulation_mode:
            # Simulate soil moisture values
            import random
            return round(30 + random.uniform(-15, 25), 1)  # 15-55%
        
        try:
            # For analog sensor, we need ADC conversion
            # This is a simplified version - you'll need proper ADC setup
            import spidev
            
            spi = spidev.SpiDev()
            spi.open(0, 0)  # SPI bus 0, device 0
            
            # Read from MCP3008 ADC (common for Raspberry Pi)
            adc_channel = 0  # Channel for soil moisture sensor
            raw_value = self._read_adc(spi, adc_channel)
            
            # Convert to percentage (0-1023 -> 0-100%)
            moisture_percent = (raw_value / 1023.0) * 100
            spi.close()
            
            return round(moisture_percent, 1)
            
        except ImportError:
            logger.error("spidev library not installed for ADC")
            return None
        except Exception as e:
            logger.error(f"Soil moisture read error: {e}")
            return None
    
    def _read_adc(self, spi, channel):
        """Read value from MCP3008 ADC"""
        # MCP3008 protocol
        cmd = 0x06  # Start bit + single-ended
        cmd <<= 5   # Shift to make room for channel
        cmd |= channel
        cmd <<= 2   # Make room for null bit
        
        resp = spi.xfer2([cmd, 0x00, 0x00])
        result = (resp[0] & 0x01) << 9
        result |= (resp[1] & 0xFF) << 1
        result |= (resp[2] & 0x80) >> 7
        
        return result
    
    def read_light_sensor(self) -> float:
        """
        Read ambient light from digital light sensor
        Returns:
            Light level in lux
        """
        if self.simulation_mode:
            # Simulate light levels
            import random
            return round(50 + random.uniform(0, 500), 1)  # 50-550 lux
        
        try:
            import smbus
            
            # I2C address for light sensor (check your sensor's datasheet)
            I2C_ADDRESS = 0x29  # Common for TSL2561
            bus = smbus.SMBus(1)  # I2C bus 1
            
            # Read light sensor (simplified - check your sensor's protocol)
            # This is a generic example for TSL2561
            data = bus.read_i2c_block_data(I2C_ADDRESS, 0x0C, 2)
            lux = (data[1] << 8) | data[0]
            
            bus.close()
            return round(lux, 1)
            
        except ImportError:
            logger.error("smbus library not installed for I2C")
            return None
        except Exception as e:
            logger.error(f"Light sensor read error: {e}")
            return None
    
    def read_water_level(self) -> Dict[str, bool]:
        """
        Read water level sensors (3 positions: top, middle, bottom)
        Returns:
            Dictionary with water level status for each position
        """
        if self.simulation_mode:
            # Simulate water levels
            import random
            return {
                "top": random.choice([True, False]),
                "middle": random.choice([True, False]),
                "bottom": random.choice([True, False])
            }
        
        try:
            water_levels = {}
            positions = ["top", "middle", "bottom"]
            
            for i, position in enumerate(positions):
                pin = GPIO_CONFIG["water_level_pins"][i]
                # Contact sensors: HIGH when water detected, LOW when dry
                water_detected = self.GPIO.input(pin) == self.GPIO.HIGH
                water_levels[position] = water_detected
            
            return water_levels
            
        except Exception as e:
            logger.error(f"Water level read error: {e}")
            return {"top": False, "middle": False, "bottom": False}
    
    def control_pump(self, pump_number: int, duration_seconds: float, water_amount_ml: float = None):
        """
        Control peristaltic pump
        Args:
            pump_number: 1 or 2 (which pump to control)
            duration_seconds: How long to run the pump
            water_amount_ml: Expected water amount (for logging)
        """
        if pump_number not in [1, 2]:
            logger.error(f"Invalid pump number: {pump_number}")
            return False
        
        pump_name = f"pump{pump_number}"
        pump_pin = GPIO_CONFIG[f"pump{pump_number}_pin"]
        
        try:
            if not self.simulation_mode:
                # Enable pump via MOSFET
                self.GPIO.output(GPIO_CONFIG["pump_enable_pin"], self.GPIO.HIGH)
                time.sleep(0.1)  # Brief delay for MOSFET activation
                
                # Turn on specific pump
                self.GPIO.output(pump_pin, self.GPIO.HIGH)
                logger.info(f"Pump {pump_number} started for {duration_seconds} seconds")
            
            # Record pump start time
            start_time = time.time()
            self.pump_status[pump_name] = True
            
            # Run pump for specified duration
            time.sleep(duration_seconds)
            
            if not self.simulation_mode:
                # Turn off pump
                self.GPIO.output(pump_pin, self.GPIO.LOW)
                self.GPIO.output(GPIO_CONFIG["pump_enable_pin"], self.GPIO.LOW)
            
            # Update status
            self.pump_status[pump_name] = False
            self.pump_status["last_watered"] = datetime.now().isoformat()
            
            # Log watering event
            event = {
                "timestamp": datetime.now().isoformat(),
                "pump": pump_number,
                "duration": duration_seconds,
                "water_amount_ml": water_amount_ml,
                "status": "completed"
            }
            
            logger.info(f"Pump {pump_number} watering completed: {event}")
            return True
            
        except Exception as e:
            logger.error(f"Pump {pump_number} control error: {e}")
            if not self.simulation_mode:
                # Ensure pumps are turned off on error
                self.GPIO.output(pump_pin, self.GPIO.LOW)
                self.GPIO.output(GPIO_CONFIG["pump_enable_pin"], self.GPIO.LOW)
            return False
    
    def read_all_sensors(self) -> Dict:
        """
        Read all sensors and return comprehensive data
        Returns:
            Dictionary with all sensor readings
        """
        timestamp = datetime.now().isoformat()
        
        # Read individual sensors
        temperature, humidity = self.read_dht22()
        soil_moisture = self.read_soil_moisture()
        light_level = self.read_light_sensor()
        water_levels = self.read_water_level()
        
        # Calculate water tank level percentage
        water_percentage = self._calculate_water_percentage(water_levels)
        
        sensor_data = {
            "timestamp": timestamp,
            "temperature_celsius": temperature,
            "humidity_percent": humidity,
            "soil_moisture_percent": soil_moisture,
            "light_lux": light_level,
            "water_levels": water_levels,
            "water_tank_percentage": water_percentage,
            "pump_status": self.pump_status.copy()
        }
        
        # Store in history
        self.sensor_history.append(sensor_data)
        self.last_sensor_readings = sensor_data
        
        # Keep only last 100 readings
        if len(self.sensor_history) > 100:
            self.sensor_history = self.sensor_history[-100:]
        
        return sensor_data
    
    def _calculate_water_percentage(self, water_levels: Dict[str, bool]) -> float:
        """Calculate water tank percentage based on sensor readings"""
        if water_levels["top"]:
            return 100.0
        elif water_levels["middle"]:
            return 66.7
        elif water_levels["bottom"]:
            return 33.3
        else:
            return 0.0
    
    def set_status_led(self, status: str):
        """Control status LED"""
        if self.simulation_mode:
            logger.info(f"Status LED: {status}")
            return
        
        try:
            if status == "normal":
                self.GPIO.output(GPIO_CONFIG["status_led"], self.GPIO.HIGH)
                self.GPIO.output(GPIO_CONFIG["warning_led"], self.GPIO.LOW)
            elif status == "warning":
                self.GPIO.output(GPIO_CONFIG["status_led"], self.GPIO.LOW)
                self.GPIO.output(GPIO_CONFIG["warning_led"], self.GPIO.HIGH)
            else:
                self.GPIO.output(GPIO_CONFIG["status_led"], self.GPIO.LOW)
                self.GPIO.output(GPIO_CONFIG["warning_led"], self.GPIO.LOW)
                
        except Exception as e:
            logger.error(f"LED control error: {e}")
    
    def get_system_status(self) -> Dict:
        """Get comprehensive system status"""
        sensor_data = self.read_all_sensors()
        
        return {
            "hardware_status": "operational" if self.gpio_initialized else "simulation",
            "simulation_mode": self.simulation_mode,
            "last_reading": sensor_data,
            "pump_status": self.pump_status,
            "sensor_history_count": len(self.sensor_history),
            "gpio_initialized": self.gpio_initialized
        }
    
    def cleanup(self):
        """Clean up GPIO resources"""
        if not self.simulation_mode and self.gpio_initialized:
            try:
                # Turn off all outputs
                output_pins = [
                    GPIO_CONFIG["pump1_pin"],
                    GPIO_CONFIG["pump2_pin"],
                    GPIO_CONFIG["pump_enable_pin"],
                    GPIO_CONFIG["status_led"],
                    GPIO_CONFIG["warning_led"]
                ]
                
                for pin in output_pins:
                    self.GPIO.output(pin, self.GPIO.LOW)
                
                self.GPIO.cleanup()
                logger.info("GPIO cleanup completed")
                
            except Exception as e:
                logger.error(f"GPIO cleanup error: {e}")

# Example usage and testing
if __name__ == "__main__":
    # Test hardware interface
    hardware = HardwareInterface(simulation_mode=True)
    
    print("=== Hardware Interface Test ===")
    print(f"Simulation mode: {hardware.simulation_mode}")
    
    # Test sensor readings
    print("\n--- Sensor Readings ---")
    sensor_data = hardware.read_all_sensors()
    for key, value in sensor_data.items():
        print(f"{key}: {value}")
    
    # Test pump control
    print("\n--- Pump Test ---")
    success = hardware.control_pump(1, 2.0, 100)
    print(f"Pump test result: {success}")
    
    # Test system status
    print("\n--- System Status ---")
    status = hardware.get_system_status()
    for key, value in status.items():
        print(f"{key}: {value}")
    
    # Cleanup
    hardware.cleanup()
    print("\nTest completed!")

File: hardware_drivers.c
#include "hardware_drivers.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <math.h>
#include <sys/time.h>
#include <stdarg.h>

// I2C and SPI includes for sensors
#include <linux/i2c-dev.h>
#include <fcntl.h>
#include <sys/ioctl.h>

// DHT22 sensor implementation
static bool dht22_read_raw(hardware_interface_t *hw, uint8_t *data) {
    if (hw->simulation_mode) {
        // Simulate realistic values
        data[0] = 0x40; // humidity high
        data[1] = 0x20; // humidity low
        data[2] = 0x20; // temperature high
        data[3] = 0x10; // temperature low
        data[4] = 0x90; // checksum
        return true;
    }
    
    // DHT22 protocol implementation
    // This is a simplified version - full implementation would require precise timing
    gpiod_line_set_value(hw->dht22_line, 0);
    usleep(18000); // 18ms low signal
    gpiod_line_set_value(hw->dht22_line, 1);
    usleep(30); // 30us high signal
    
    // Read data bits (simplified)
    // In a real implementation, you'd need precise timing for each bit
    for (int i = 0; i < 5; i++) {
        data[i] = 0;
        for (int j = 0; j < 8; j++) {
            usleep(50); // Wait for bit
            int bit = gpiod_line_get_value(hw->dht22_line);
            data[i] |= (bit << (7 - j));
            usleep(70); // Wait for next bit
        }
    }
    
    // Verify checksum
    uint8_t checksum = data[0] + data[1] + data[2] + data[3];
    return (checksum == data[4]);
}

// ADC reading for soil moisture (MCP3008)
static uint16_t read_adc_channel(int spi_fd, int channel) {
    if (spi_fd < 0) return 0;
    
    uint8_t tx[3] = {0x01, (0x08 | channel) << 4, 0x00};
    uint8_t rx[3];
    
    struct spi_ioc_transfer tr = {
        .tx_buf = (unsigned long)tx,
        .rx_buf = (unsigned long)rx,
        .len = 3,
        .speed_hz = 1000000,
        .delay_usecs = 0,
        .bits_per_word = 8,
    };
    
    if (ioctl(spi_fd, SPI_IOC_MESSAGE(1), &tr) < 0) {
        return 0;
    }
    
    return ((rx[1] & 0x03) << 8) | rx[2];
}

// I2C light sensor reading (TSL2561)
static float read_tsl2561_lux(int i2c_fd) {
    if (i2c_fd < 0) return 0.0;
    
    // Enable sensor
    uint8_t enable_cmd = 0x03;
    if (write(i2c_fd, &enable_cmd, 1) != 1) return 0.0;
    
    usleep(100000); // Wait 100ms for integration
    
    // Read channel 0 (visible + IR)
    uint8_t cmd0 = 0x0C;
    if (write(i2c_fd, &cmd0, 1) != 1) return 0.0;
    
    uint8_t data[2];
    if (read(i2c_fd, data, 2) != 2) return 0.0;
    
    uint16_t ch0 = (data[1] << 8) | data[0];
    
    // Read channel 1 (IR only)
    uint8_t cmd1 = 0x0E;
    if (write(i2c_fd, &cmd1, 1) != 1) return 0.0;
    
    if (read(i2c_fd, data, 2) != 2) return 0.0;
    
    uint16_t ch1 = (data[1] << 8) | data[0];
    
    // Calculate lux (simplified formula)
    if (ch0 == 0) return 0.0;
    
    float ratio = (float)ch1 / ch0;
    float lux = 0.0;
    
    if (ratio <= 0.50) {
        lux = 0.0304 * ch0 - 0.062 * ch0 * pow(ratio, 1.4);
    } else if (ratio <= 0.61) {
        lux = 0.0224 * ch0 - 0.031 * ch1;
    } else if (ratio <= 0.80) {
        lux = 0.0128 * ch0 - 0.0153 * ch1;
    } else if (ratio <= 1.30) {
        lux = 0.00146 * ch0 - 0.00112 * ch1;
    } else {
        lux = 0.0;
    }
    
    return lux;
}

hardware_interface_t* hardware_init(bool simulation_mode) {
    hardware_interface_t *hw = malloc(sizeof(hardware_interface_t));
    if (!hw) {
        hardware_log("ERROR", "Failed to allocate memory for hardware interface");
        return NULL;
    }
    
    memset(hw, 0, sizeof(hardware_interface_t));
    hw->simulation_mode = simulation_mode;
    
    if (!simulation_mode) {
        // Open GPIO chip
        hw->chip = gpiod_chip_open_by_name("gpiochip0");
        if (!hw->chip) {
            hardware_log("ERROR", "Failed to open GPIO chip: %s", strerror(errno));
            free(hw);
            return NULL;
        }
        
        // Request GPIO lines
        hw->dht22_line = gpiod_chip_get_line(hw->chip, DHT22_PIN);
        hw->soil_moisture_line = gpiod_chip_get_line(hw->chip, SOIL_MOISTURE_PIN);
        hw->water_level_lines[0] = gpiod_chip_get_line(hw->chip, 5);
        hw->water_level_lines[1] = gpiod_chip_get_line(hw->chip, 6);
        hw->water_level_lines[2] = gpiod_chip_get_line(hw->chip, 7);
        hw->pump1_line = gpiod_chip_get_line(hw->chip, PUMP1_PIN);
        hw->pump2_line = gpiod_chip_get_line(hw->chip, PUMP2_PIN);
        hw->pump_enable_line = gpiod_chip_get_line(hw->chip, PUMP_ENABLE_PIN);
        hw->status_led_line = gpiod_chip_get_line(hw->chip, STATUS_LED_PIN);
        hw->warning_led_line = gpiod_chip_get_line(hw->chip, WARNING_LED_PIN);
        
        // Configure input lines
        if (gpiod_line_request_input(hw->dht22_line, "automated_planter") < 0) {
            hardware_log("ERROR", "Failed to request DHT22 line");
            goto cleanup;
        }
        
        if (gpiod_line_request_input(hw->soil_moisture_line, "automated_planter") < 0) {
            hardware_log("ERROR", "Failed to request soil moisture line");
            goto cleanup;
        }
        
        for (int i = 0; i < 3; i++) {
            if (gpiod_line_request_input(hw->water_level_lines[i], "automated_planter") < 0) {
                hardware_log("ERROR", "Failed to request water level line %d", i);
                goto cleanup;
            }
        }
        
        // Configure output lines
        if (gpiod_line_request_output(hw->pump1_line, "automated_planter", 0) < 0) {
            hardware_log("ERROR", "Failed to request pump1 line");
            goto cleanup;
        }
        
        if (gpiod_line_request_output(hw->pump2_line, "automated_planter", 0) < 0) {
            hardware_log("ERROR", "Failed to request pump2 line");
            goto cleanup;
        }
        
        if (gpiod_line_request_output(hw->pump_enable_line, "automated_planter", 0) < 0) {
            hardware_log("ERROR", "Failed to request pump enable line");
            goto cleanup;
        }
        
        if (gpiod_line_request_output(hw->status_led_line, "automated_planter", 0) < 0) {
            hardware_log("ERROR", "Failed to request status LED line");
            goto cleanup;
        }
        
        if (gpiod_line_request_output(hw->warning_led_line, "automated_planter", 0) < 0) {
            hardware_log("ERROR", "Failed to request warning LED line");
            goto cleanup;
        }
        
        hw->gpio_initialized = true;
        hardware_log("INFO", "Hardware interface initialized successfully");
    } else {
        hardware_log("INFO", "Hardware interface initialized in simulation mode");
    }
    
    return hw;
    
cleanup:
    hardware_cleanup(hw);
    return NULL;
}

void hardware_cleanup(hardware_interface_t *hw) {
    if (!hw) return;
    
    if (hw->gpio_initialized) {
        // Turn off all outputs
        if (hw->pump1_line) gpiod_line_set_value(hw->pump1_line, 0);
        if (hw->pump2_line) gpiod_line_set_value(hw->pump2_line, 0);
        if (hw->pump_enable_line) gpiod_line_set_value(hw->pump_enable_line, 0);
        if (hw->status_led_line) gpiod_line_set_value(hw->status_led_line, 0);
        if (hw->warning_led_line) gpiod_line_set_value(hw->warning_led_line, 0);
        
        // Release lines
        if (hw->dht22_line) gpiod_line_release(hw->dht22_line);
        if (hw->soil_moisture_line) gpiod_line_release(hw->soil_moisture_line);
        for (int i = 0; i < 3; i++) {
            if (hw->water_level_lines[i]) gpiod_line_release(hw->water_level_lines[i]);
        }
        if (hw->pump1_line) gpiod_line_release(hw->pump1_line);
        if (hw->pump2_line) gpiod_line_release(hw->pump2_line);
        if (hw->pump_enable_line) gpiod_line_release(hw->pump_enable_line);
        if (hw->status_led_line) gpiod_line_release(hw->status_led_line);
        if (hw->warning_led_line) gpiod_line_release(hw->warning_led_line);
        
        // Close chip
        if (hw->chip) gpiod_chip_close(hw->chip);
        
        hardware_log("INFO", "GPIO cleanup completed");
    }
    
    free(hw);
}

bool read_dht22(hardware_interface_t *hw, float *temperature, float *humidity) {
    if (!hw || !temperature || !humidity) return false;
    
    uint8_t data[5];
    
    if (!dht22_read_raw(hw, data)) {
        hardware_log("WARNING", "DHT22 sensor read failed");
        return false;
    }
    
    // Convert raw data to temperature and humidity
    *humidity = ((data[0] << 8) | data[1]) / 10.0;
    *temperature = (((data[2] & 0x7F) << 8) | data[3]) / 10.0;
    
    // Handle negative temperature
    if (data[2] & 0x80) {
        *temperature = -(*temperature);
    }
    
    return true;
}

float read_soil_moisture(hardware_interface_t *hw) {
    if (!hw) return 0.0;
    
    if (hw->simulation_mode) {
        // Simulate soil moisture values
        return 30.0 + (rand() % 400) / 10.0; // 30.0-70.0%
    }
    
    // Open SPI device for ADC
    int spi_fd = open("/dev/spidev0.0", O_RDWR);
    if (spi_fd < 0) {
        hardware_log("ERROR", "Failed to open SPI device: %s", strerror(errno));
        return 0.0;
    }
    
    uint16_t raw_value = read_adc_channel(spi_fd, 0);
    close(spi_fd);
    
    // Convert to percentage (0-1023 -> 0-100%)
    float moisture_percent = (raw_value / 1023.0) * 100.0;
    
    return moisture_percent;
}

float read_light_sensor(hardware_interface_t *hw) {
    if (!hw) return 0.0;
    
    if (hw->simulation_mode) {
        // Simulate light levels
        return 50.0 + (rand() % 5000) / 10.0; // 50.0-550.0 lux
    }
    
    // Open I2C device
    int i2c_fd = open("/dev/i2c-1", O_RDWR);
    if (i2c_fd < 0) {
        hardware_log("ERROR", "Failed to open I2C device: %s", strerror(errno));
        return 0.0;
    }
    
    // Set I2C slave address
    if (ioctl(i2c_fd, I2C_SLAVE, 0x29) < 0) {
        hardware_log("ERROR", "Failed to set I2C slave address: %s", strerror(errno));
        close(i2c_fd);
        return 0.0;
    }
    
    float lux = read_tsl2561_lux(i2c_fd);
    close(i2c_fd);
    
    return lux;
}

void read_water_level(hardware_interface_t *hw, bool *top, bool *middle, bool *bottom) {
    if (!hw || !top || !middle || !bottom) return;
    
    if (hw->simulation_mode) {
        // Simulate water levels
        *top = (rand() % 2) == 1;
        *middle = (rand() % 2) == 1;
        *bottom = (rand() % 2) == 1;
        return;
    }
    
    if (!hw->gpio_initialized) return;
    
    // Read water level sensors (HIGH when water detected, LOW when dry)
    *top = gpiod_line_get_value(hw->water_level_lines[0]) == 1;
    *middle = gpiod_line_get_value(hw->water_level_lines[1]) == 1;
    *bottom = gpiod_line_get_value(hw->water_level_lines[2]) == 1;
}

bool control_pump(hardware_interface_t *hw, int pump_number, float duration_seconds, float water_amount_ml) {
    if (!hw || pump_number < 1 || pump_number > 2) {
        hardware_log("ERROR", "Invalid pump number: %d", pump_number);
        return false;
    }
    
    struct gpiod_line *pump_line = (pump_number == 1) ? hw->pump1_line : hw->pump2_line;
    
    if (!hw->simulation_mode && hw->gpio_initialized) {
        // Enable pump via MOSFET
        gpiod_line_set_value(hw->pump_enable_line, 1);
        usleep(100000); // 100ms delay for MOSFET activation
        
        // Turn on specific pump
        gpiod_line_set_value(pump_line, 1);
        hardware_log("INFO", "Pump %d started for %.1f seconds", pump_number, duration_seconds);
    }
    
    // Record pump start time
    hw->pump_status.last_watered = time(NULL);
    if (pump_number == 1) {
        hw->pump_status.pump1_active = true;
    } else {
        hw->pump_status.pump2_active = true;
    }
    
    // Run pump for specified duration
    usleep((int)(duration_seconds * 1000000));
    
    if (!hw->simulation_mode && hw->gpio_initialized) {
        // Turn off pump
        gpiod_line_set_value(pump_line, 0);
        gpiod_line_set_value(hw->pump_enable_line, 0);
    }
    
    // Update status
    if (pump_number == 1) {
        hw->pump_status.pump1_active = false;
    } else {
        hw->pump_status.pump2_active = false;
    }
    
    hardware_log("INFO", "Pump %d watering completed", pump_number);
    return true;
}

void set_status_led(hardware_interface_t *hw, const char *status) {
    if (!hw || !status) return;
    
    if (hw->simulation_mode) {
        hardware_log("INFO", "Status LED: %s", status);
        return;
    }
    
    if (!hw->gpio_initialized) return;
    
    if (strcmp(status, "normal") == 0) {
        gpiod_line_set_value(hw->status_led_line, 1);
        gpiod_line_set_value(hw->warning_led_line, 0);
    } else if (strcmp(status, "warning") == 0) {
        gpiod_line_set_value(hw->status_led_line, 0);
        gpiod_line_set_value(hw->warning_led_line, 1);
    } else {
        gpiod_line_set_value(hw->status_led_line, 0);
        gpiod_line_set_value(hw->warning_led_line, 0);
    }
}

sensor_data_t read_all_sensors(hardware_interface_t *hw) {
    sensor_data_t data = {0};
    
    if (!hw) return data;
    
    data.timestamp = time(NULL);
    
    // Read individual sensors
    read_dht22(hw, &data.temperature_celsius, &data.humidity_percent);
    data.soil_moisture_percent = read_soil_moisture(hw);
    data.light_lux = read_light_sensor(hw);
    read_water_level(hw, &data.water_level_top, &data.water_level_middle, &data.water_level_bottom);
    
    // Calculate water tank level percentage
    data.water_tank_percentage = calculate_water_percentage(
        data.water_level_top, 
        data.water_level_middle, 
        data.water_level_bottom
    );
    
    return data;
}

float calculate_water_percentage(bool top, bool middle, bool bottom) {
    if (top) return 100.0;
    if (middle) return 66.7;
    if (bottom) return 33.3;
    return 0.0;
}

void hardware_log(const char *level, const char *format, ...) {
    time_t now = time(NULL);
    struct tm *tm_info = localtime(&now);
    char timestamp[64];
    strftime(timestamp, sizeof(timestamp), "%Y-%m-%d %H:%M:%S", tm_info);
    
    printf("[%s] [%s] ", timestamp, level);
    
    va_list args;
    va_start(args, format);
    vprintf(format, args);
    va_end(args);
    
    printf("\n");
}

time_t get_current_timestamp(void) {
    return time(NULL);
}
